pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -183,6 +194,10 @@ public void fetchStringValues(String column, int[] inDocIds, int length, String[
     } else {
       _singleValueSetMap.get(column).getStringValues(inDocIds, 0, length, outValues, 0);
     }
+    // todo: optimize!
+    for( String val : outValues) {
```

Comment:
This can be really expensive for low latency/high throughput use cases. You can optimize it by computing the length inside of SingleValueSet.getStringValues, since it is looping over the strings anyways.